docs/internal/3i.md

```diff
@@ -0,0 +1,141 @@
+
+**Why not interposable?**
+At the time `trigger_harsh_cage_exit` or `harsh_cage_exit` is invoked, the target Cage or Grate is assumed to have unreliable memory and control flow. During the execution of these calls, the syscall table of the target cage/grate is either being torn down or may already be corrupted, meaning the call path itself is no longer trustworthy.
+
+The cleanup process must complete system-level invariants such as: unmapping vmmap regions, cleaning fdtables, waking waiters, canceling schedulers or timers, etc.. Allowing these calls to be interposable would permit third-party grates to inject arbitrary logic (e.g., blocking, allocation, or reentrancy), which could stall or disrupt the teardown sequence, resulting in resource leaks, deadlocks, or zombie cages/grates.
+
+## Grates
+
+### Overview
+
+Grates are cages that provide services to descendant cages by interposing on their system calls. They can execute trusted operations (e.g., syscall filtering, tracing, or in-memory file I/O) on behalf of other cages without requiring changes to the microvisor’s trusted computing base (TCB).
+
+From 3i’s perspective, a grate is both:
+
+- A consumer of system calls from descendant cages, and
+- A service provider exposing its own callable entry points to the 3i runtime.
+
+### Design Goals
+
+3i’s public API is runtime-agnostic, allowing it to be adapted to multiple backends (e.g., Wasmtime, MPK, or future runtimes).
+
+Runtime-specific mechanisms, such as Wasmtime’s `VMContext` management, are implemented in the corresponding runtime layer (e.g., the Wasmtime adapter). These runtime hooks are exposed to 3i as C-ABI function pointers and registered in 3i’s global table, decoupling runtime details from 3i’s core logic.
+
+### Wasm Implementation
+
+Grate execution relies on a re-entry mechanism that bridges Wasm runtime contexts (e.g., Wasmtime instances) and 3i’s routing layer. 
+
+Each Wasm module in Wasmtime has its own `VMContext`, which encapsulates all runtime state required for that instance’s execution. When 3i performs a cross-module call (e.g., from _cage A_ to a _grate B_), the currently executing context belongs to _A_, but 3i needs to restore the latest runtime state of _B_ before re-entering its Wasm code. 
+
+Because Rust’s lifetime system prevents us from directly sharing or globally storing references to the caller’s runtime structures, we instead store and pass raw pointers to the target module’s `VMContext`. 3i uses a registered callback function pointer and retrived the corresponding runtime context (`VMContext`) from pointer to re-enter the target Wasm instance. This allows 3i to reconstruct the correct execution environment for _B_ at re-entry time, ensuring cross-instance transitions without violating Rust’s ownership and lifetime rules.
```

Comment:
I find this confusing.  I don't really understand.  Is there a location where this information is stored that persists?  Do we just do a table lookup for this?

Reply:
I do still need some help here.  Is this really a Rust problem?  Isn't this that you need to restore the state and you don't get passed what you need lower in the call stack?